### ISSUE TITLE
Expo Updates Not working With Expo Custom-Dev-Client and EAS Build

### ISSUE THREAD

**Issue by vipinjoshiMeta** [expo/expo#15978]
### Summary

I'm trying to integrate expo-updates in a Expo Custom Dev Client bare managed project.
For some how i can be able to build apk with eas build & its working fine.

But Expo updates not working
i'm publishing update using **expo publish** command

**expo publish --release-channel channel-name**

But in the app i'm not getting any update notification than i try to setup config for update manually

`const triggerUpdateCheck = async () => {

        try {
            const update = Updates.checkForUpdateAsync();
            if (update.isAvailable) {
                await Updates.fetchUpdateAsync()
                alert("An update is available, Restart your app to see it.")
            } else {
                alert("No update is available")
            }
        } catch (err) {
            alert(error.message)
        }
    }`
    
    But i'm always getting there is no update available
    I also had cross check both apk & publish update release channels.
    
    For the manual update what i have done in a app.json
    
    `"updates": {
      "enabled": true,
      "checkAutomatically": "ON_ERROR_RECOVERY",
      "fallbackToCacheTimeout": 0
    },`

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### Environment

Expo CLI 5.0.3 environment info:
    System:
      OS: Windows 10 10.0.19043
    Binaries:
      Node: 14.15.4 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.11 - ~\AppData\Roaming\npm\yarn.CMD
      npm: 7.24.0 - C:\Program Files\nodejs\npm.CMD
    SDKs:
      Android SDK:
        API Levels: 23, 26, 27, 28, 29, 30, 31
        Build Tools: 23.0.1, 28.0.3, 29.0.2, 29.0.3, 30.0.2, 31.0.0, 32.0.0
        System Images: android-30 | Google APIs Intel x86 Atom, android-30 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: Version     2020.3.0.0 AI-203.7717.56.2031.7583922
    npmPackages:
      expo: ~43.0.2 => 43.0.3
      react: 17.0.1 => 17.0.1
      react-dom: 17.0.1 => 17.0.1
      react-native: 0.64.3 => 0.64.3
      react-native-web: 0.17.1 => 0.17.1
      react-navigation: ^4.4.4 => 4.4.4
    Expo Workflow: bare

### Reproducible demo

`const triggerUpdateCheck = async () => {
        try {
            const update = Updates.checkForUpdateAsync();
            if (update.isAvailable) {
                await Updates.fetchUpdateAsync()
                alert("An update is available, Restart your app to see it.")
            } else {
                alert("No update is available")
            }
        } catch (err) {
            alert(error.message)
        }
    }
`

**App.json file changes:**

```
"updates": {
      "enabled": true,
      "checkAutomatically": "ON_ERROR_RECOVERY",
      "fallbackToCacheTimeout": 0
    }
```
    
    
    **EAS Config**
    
   ```
 {
  "cli": {
    "version": ">= 0.38.2"
  },
  "build": {
    "development": {
      "developmentClient": true,
      "distribution": "internal"
    },
    "preview": {
      "distribution": "internal"
    },
    "production": {
      "android": {
        "buildType": "apk",
        "gradleCommand": ":app:assembleRelease"
      },
      "releaseChannel": "production",
      "env": {
        "APP_ENV": "production"
      }
    },
    "production-debug": {
      "android": {
        "buildType": "apk",
        "gradleCommand": ":app:assembleDebug"
      },
      "releaseChannel": "production",
      "env": {
        "APP_ENV": "production"
      }
    },
    "test": {
      "android": {
        "buildType": "apk",
        "gradleCommand": ":app:assembleRelease"
      },
      "releaseChannel": "test",
      "env": {
        "APP_ENV": "staging"
      }
    },
    "test-debug": {
      "android": {
        "buildType": "apk",
        "gradleCommand": ":app:assembleDebug"
      },
      "releaseChannel": "test",
      "env": {
        "APP_ENV": "staging"
      }
    },
    "staging": {
      "android": {
        "buildType": "apk",
        "gradleCommand": ":app:assembleRelease"
      },
      "releaseChannel": "staging",
      "env": {
        "APP_ENV": "staging"
      }
    },
    "staging-debug": {
      "android": {
        "buildType": "apk",
        "gradleCommand": ":app:assembleDebug"
      },
      "releaseChannel": "staging",
      "env": {
        "APP_ENV": "staging"
      }
    },
    "production-bundle": {}
  },
  "submit": {
    "production": {}
  }
}

```

Will be a great help if anyone can give some guidelines, how it should be implemented in a better way.
Basically we need some automatically update check feature on the app start & a mechanism also to give some interface to manually update the app.

**Comment by tcdavis** [comment]
Thanks for the report.  
If I understand the issue is that the update.isAvailable is always false, even after you publish a new update?
What steps are you taking to load the project in your dev client?

**Comment by vipinjoshiMeta** [comment]
Well In our project we are using **"expo-updates": "~0.10.13",**

Firstly i was trying to setup auto update i read docs to set these properties in  a app.json file.
`"updates": {
      "enabled": true,
      "checkAutomatically": "ON_LOAD",
      "fallbackToCacheTimeout": 60000
    },`

 **Doing these changes in a app.js (root component of app)**
'/*React.useEffect(() => {
        console.log(TAG, 'useEffectReactToUpdates')
        reactToUpdates()
    }, [])

    const reactToUpdates = async (): void => {
        Updates.addListener((event) => {
           if (event.type === UpdateEventType.UPDATE_AVAILABLE) {
                console.log(TAG, "An update is available, Restart your app to see it.")
                Alert.alert("An update is available, Restart your app to see it.")
                // Updates.reloadAsync() // will restart the app
            }
        })
    }*/'



than I try to publish the update by incrementing version 1.0.0 to 1.0.1
`"version": "1.0.1",`

than i also try to publish by incrementing both version & version code 
`"android": {
      "adaptiveIcon": {
        "foregroundImage": "./assets/adaptive-icon.png",
        "backgroundColor": "#FFFFFF"
      },
      "permissions": [
        "android.permission.INTERNET"
      ],
      "package": "test-ota",
      "versionCode": 2
    },`

But we doesn't see any effect on pre generated apk both times, I do restart the app many times but nothing happened.
Than I try to setup manual - 
For the manual update check,  I just change these properties & try same steps but nothing goes happens.

`"updates": {
      "enabled": true,
      "checkAutomatically": "ON_ERROR_RECOVERY",
      "fallbackToCacheTimeout": 0
    },`

& For the Manual Update I'm just triggering this function on a click of a button

`const triggerUpdateCheck = async () => {
        try {
            const update = await Updates.checkForUpdateAsync();
            if (update.isAvailable) {
                await Updates.fetchUpdateAsync()
                alert("An update is available, Restart your app to see it.")
            } else {
                alert("No update is available")
            }
        } catch (err) {
            alert(error.message)
        }
    }`

`<AppButton
                        type="primary"
                        height={40}
                        margin={8}
                        onPress={async () => {
                            setUpdateChecking(true)
                            await triggerUpdateCheck()
                            setUpdateChecking(false)
                        }}
                        title={'Check For Update'}
                    />`

**Comment by vipinjoshiMeta** [comment]
@tcdavis after running some debugging I've console output some errors logs Like

**Failed to download manifest from URL**

Can you help me to figure-out why is it keep giving me this kind of error what else it needs.
or some guidelines how to properly push an update to an existing apk.

Should I've  to run **expo run:android** each before publishing any update?

**Comment by LinFeiLong** [comment]
Hi @vipinjoshiMeta ,
I'm facing the same issue.

**Comment by LinFeiLong** [comment]
@vipinjoshiMeta I resolved the issue on my side.
It's not an issue properly talking.
I had to add 
`"runtimeVersion": {
      "policy": "nativeVersion"
    },`
to the app.json.
Then the automatic update worked.

**Comment by vipinjoshiMeta** [comment]
@tcdavis @LinFeiLong thanks for your time.

its working now.🏳

But in my case there was something else whenever I run **expo run:android**
**The Android manifest not getting updating.**

So these property should match with app.json configuration, so i had to delete & recreate android directory whenever I need to do some change in configuration & reconfigure all the dependencies.

```
<meta-data android:name="expo.modules.updates.ENABLED" android:value="true"/>
<meta-data android:name="expo.modules.updates.EXPO_SDK_VERSION" andaw3seeroid:value="43.0.0"/>
<meta-data android:name="expo.modules.updates.EXPO_UPDATES_CHECK_ON_LAUNCH" android:value="NEVER"/>
<meta-data android:name="expo.modules.updates.EXPO_UPDATES_LAUNCH_WAIT_MS" android:value="0"/>
<meta-data android:name="expo.modules.updates.EXPO_UPDATE_URL" android:value="https://exp.host/@URL"/>
```

**Comment by LinFeiLong** [comment]
@vipinjoshiMeta You should probably close your issue. Thx.
